### ISSUE TITLE
Issues with data being sent from winpty@0.4.1

### ISSUE THREAD

**Issue by Tyriar** [rprichard/winpty#96]
After hooking up v0.4.1 to [Tyriar/node-pty](https://github.com/Tyriar/node-pty), a few issues have been fixed (arrows in WSL https://github.com/Microsoft/vscode/issues/10163, prompt getting out of sync in git bash when backspace is pressed https://github.com/Microsoft/vscode/issues/7345).

However I've found a few regressions that make me worry about taking the uplevel, namely:

- In Git Bash (`C:\Program Files\Git\usr\bin\bash.exe`), instead of the prompt getting out of sync, now the terminal emulator is told to move the cursor 1 position to the *right*
  
  ![image](https://cloud.githubusercontent.com/assets/2193314/21960156/aa030050-da94-11e6-8d4b-f808bf9ff859.png)

  After pressing backspace, winpty sends `\x1B[5G`, resulting in:

  ![image](https://cloud.githubusercontent.com/assets/2193314/21960159/d7e0ef78-da94-11e6-9a53-7a3873fee731.png)

- In WSL and Git Bash, the arrow keys don't seem to function within vim. The terminal emulator sends `\x1B[D` to winpty and I get a bell sound and vim tells me `E388: Couldn't find definition`:

  ![image](https://cloud.githubusercontent.com/assets/2193314/21960163/43de495a-da95-11e6-9cf9-5d88a4fcf888.png)

I'm using `WINPTY_SHOW_CONSOLE=1` to show the console. The backspace issue shows if I switch to the winpty window and hit backspace, the arrows in vim seem to function fine however.

**Comment by rprichard** [comment]
The issue with cursor keys in WSL vim is probably https://github.com/rprichard/winpty/issues/90.  I think I know how to fix it, so I probably should make the fix.

The issue with cursor keys in Cygwin/MSYS vim is probably the result of node-pty setting TERM to xterm by default, when it should instead leave it unset.  (See https://github.com/rprichard/winpty/issues/43, https://youtrack.jetbrains.com/issue/IDEA-155071, https://bugs.eclipse.org/bugs/show_bug.cgi?id=496291.)  Basically, it doesn't matter that VSCode has an xterm-compatible terminal -- the issue is whether the programs attached to the console see an xterm-like environment, and they don't.  Instead, they see the Cygwin/MSYS console<->pty translation layer, which is identified as `TERM=cygwin`.  VSCode doesn't need to set TERM=cygwin, though, because Cygwin will do that automatically.

I haven't reproduced the backspace issue yet.

**Comment by Tyriar** [comment]
That would be awesome if you get a chance to fix #90.

vim was actually completely broken in git bash and the reason was because `TERM=bash.exe`, setting `TERM=xterm` allowed me to enter vim. I'll fiddle with `TERM` and see whether it impacts the backspace issue. Thanks :smiley:

**Comment by Tyriar** [comment]
Awesome, setting `TERM=xterm` after the terminal has been launched fixes it! I'll follow that up in https://github.com/Tyriar/node-pty/issues/22

**Comment by rprichard** [comment]
For git bash, the appropriate TERM would be `cygwin` rather than `xterm`.  (But because winpty works by creating an ordinary Windows console, the best thing to do is to leave TERM unset.)

Another program to test is `less`.  If I run git bash's less in an ordinary Windows console, but set TERM to xterm first, then the arrow keys won't work.  If I set TERM to cygwin (or leave TERM unset), then they work.

**Comment by Tyriar** [comment]
Yes unsetting it makes it cygwin which seems to work great (including `less`) https://github.com/Tyriar/node-pty/commit/11fc06a13a763279e6bcf38e2d8ecf9c7e591bf3

**Comment by rprichard** [comment]
Yeah, that change looks good.

I'm trying to reproduce the issues by running VSCode with node-pty's winpty_upgrade_0_4_0 branch, but I'm having trouble.  I've successfully checked out VSCode, and it runs, but when I edit package.json, it doesn't seem to have an effect.  Do you know what I'm doing wrong?  Here's my current package.json edit:

```diff
C:\rprichard\proj\vscode>git diff
diff --git a/package.json b/package.json
index 3b4dca5..5ecbe7e 100644
--- a/package.json
+++ b/package.json
@@ -30,7 +30,7 @@
     "iconv-lite": "0.4.15",
     "minimist": "1.2.0",
     "native-keymap": "0.3.0",
-    "node-pty": "0.4.1",
+    "node-pty": "file:../node-pty",
     "semver": "4.3.6",
     "vscode-debugprotocol": "1.15.0",
     "vscode-textmate": "3.1.0",
@@ -116,4 +116,4 @@
     "windows-mutex": "^0.2.0",
     "fsevents": "0.3.8"
   }
-}
\ No newline at end of file
+}
warning: LF will be replaced by CRLF in src/vs/monaco.d.ts.
The file will have its original line endings in your working directory.
```

Then I rerun `scripts\npm install` and `scripts\code.bat`.  The npm install doesn't seem to update the winpty files in `node_modules`.

**Comment by Tyriar** [comment]
If you want to build you own them you need to change npm-shrinkwrap.json, don't worry about package.json as shrinkwrap overrides the dep. Here's how you do it:

1. Remove `node_modules/node-pty`
2. Change the `resolved` property under `node-pty` in `npm-shrinkwrap.json` to git+https://github.com/Tyriar/node-pty.git#11fc06a13a763279e6bcf38e2d8ecf9c7e591bf3 (this is pointing to the typescript branch which is the latest)
3. `./scripts/npm.bat install`
4. `./scripts/code.bat`

If you need to fiddle with the TS files you will need to run `npm run tsc` to compile them (or just edit the js files).

**Comment by Tyriar** [comment]
You may be able to have npm-shrinkwrap.json resolve to relative paths, I'm not sure if you can though.

**Comment by Tyriar** [comment]
Oh and when you get it built you can open the terminal with ctrl+\` and can change the shell it's launched with by opening `settings.json` (ctrl+shift+p, "settings") and adding:

```js
"terminal.integrated.shell.windows": "C:\\path\\to\\executable",
```

**Comment by rprichard** [comment]
Thanks, it looks like it used the new node-pty code this time.  When I press "Ctrl-`", though, I just see this error:
```
The terminal process terminated with exit code: 1
Press any key to close the terminal
```

**Comment by Tyriar** [comment]
Hmm, are you on the latest master? That probably means that `node-pty` died for some reason.

One thing I realized is that the compiled js isn't checked in to the repo which is likely the problem. I forgot I was publishing beta builds to npm while testing this, change the resolved part to `https://registry.npmjs.org/node-pty/-/node-pty-0.5.0-beta3.tgz` for a recent build (with js) that you should be able to repro the problems on.

**Comment by rprichard** [comment]
Yeah, I'm using the latest master of vscode.

I changed the resolved to point to an absolute local path:
```
"resolved": "file:C:/rprichard/proj/node-pty",
```
I had the winpty_upgrade_0_4_0 branch checked out, and it worked, and the backspace issue reproduced.  TERM was set to `cmd.exe`. (or `powershell.exe`).  Once I unset TERM, the git bash backspace issue went away.

I'll try changing to your resolved setting, but I *think* there aren't any VSCode-specific things I need to examine for the moment.

**Comment by Tyriar** [comment]
Yes I think that branch illustrated the issue, you will probably need to add a setting to launch WSL though:

```js
"terminal.integrated.shell.windows": "C:\\Windows\\sysnative\\bash.exe"
```

**Comment by rprichard** [comment]
I changed the resolved to `https://registry.npmjs.org/node-pty/-/node-pty-0.5.0-beta3.tgz`, and it demonstrates the various issues (and leaves TERM set to the basename of the shell program, e.g. powershell.exe).

**Comment by Tyriar** [comment]
I haven't republished the beta build yet so that will repro, I've confirmed that one is fixed 👍

**Comment by rprichard** [comment]
This issue was fixed by https://github.com/rprichard/winpty/commit/40d5b72c43165875724825fab5b5e5303fa5c3c3.  The fix is included in today's 0.4.2 release.
